sockeye/optimizers.py

```diff
@@ -0,0 +1,244 @@
+# Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not
+# use this file except in compliance with the License. A copy of the License
+# is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed on
+# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+"""
+Extra optimizers not included in MXNet.
+"""
+
+from abc import abstractmethod
+from collections import namedtuple
+import math
+from typing import Optional, Tuple
+
+from mxnet.ndarray import NDArray, sqrt, zeros_like
+from mxnet.optimizer import Optimizer, clip
+
+from sockeye.utils import check_condition
+
+BatchState = namedtuple("BatchState", ["metric_val"])
+CheckpointState = namedtuple("CheckpointState", ["checkpoint", "metric_val"])
+
+
+class SockeyeOptimizer(Optimizer):
+    """
+    Optimizer that has access to additional information from the last batch and the last chekpoint
+    when updating weights.
+
+    :param request_optimized_metric: Whether to request the optimized metric (e.g. perplexity) in
+                                     place of optimizer loss (e.g. cross-entropy).
+    """
+    def __init__(self, request_optimized_metric: bool = False, **kwargs) -> None:
+        self.request_optimized_metric = request_optimized_metric
+        self.batch_state = Optional[BatchState]
```

Comment:
`Optional` is meant to be used as a type annotation, but shouldn't be instantiated and assigned. If you want to declare the type of the member variable you can use:
```
self.batch_size = None # type: Optional[BatchState]
```
FYI: [PEP 526](https://www.python.org/dev/peps/pep-0526/) will allow for variable annotations, but requires python 3.6 and we currently support >=3.4:
```
self.batch_size: Optional[BatchState] = None
```

Optional means a variable might be `None`, so these variables should just be initialized to `None` here.

Reply:
Thanks!  Felix and I were going back and forth looking for something that would make all versions of Python happy.